pkg/matrix/matrix.go

```diff
@@ -0,0 +1,56 @@
+/*
+Copyright 2022 The Tekton Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package matrix
+
+import (
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"strconv"
+)
+
+// FanOut produces combinations of Parameters of type String from a slice of Parameters of type Array.
+func FanOut(params []v1beta1.Param) Combinations {
+	var combinations []*Combination
+	for _, parameter := range params {
+		combinations = distributeParameter(parameter, combinations)
+	}
+	return combinations
+}
+
+func distributeParameter(param v1beta1.Param, existingCombinations []*Combination) []*Combination {
+	var expandedCombinations []*Combination
+	if len(existingCombinations) == 0 {
```

Comment:
It might be handy to have a comment explaining why the behavior is different depending on whether there already are existing combinations.

Reply:
@abayer added comments explaining the logic

Reply:
Thanks!